tests/nvme_copy_test.py

```diff
@@ -4,56 +4,124 @@
+                print("Running command:", set_features_cmd)
+                proc = subprocess.Popen(set_features_cmd,
+                                        shell=True,
+                                        stdout=subprocess.PIPE,
+                                        stdin=subprocess.PIPE)
+                proc.communicate(input=data)
+        get_ns_id_cmd = f"nvme get-ns-id {self.ns1}"
+        print("Running command:", get_ns_id_cmd)
+        proc = subprocess.Popen(get_ns_id_cmd, shell=True, stdout=subprocess.PIPE)
+        self.ns1_nsid = int(proc.stdout.read().decode().strip().split(':')[-1])
         self.setup_log_dir(self.__class__.__name__)
 
     def tearDown(self):
         """ Post Section for TestNVMeCopy """
+        print("Tearing down test...")
+        if self.host_behavior_data:
+            # restore saved host behavior support data
+            set_features_cmd = "nvme set-feature " + self.ctrl + \
+                    " --feature-id=0x16 -l 512"
+            print("Running command:", set_features_cmd)
+            proc = subprocess.Popen(set_features_cmd,
+                                    shell=True,
+                                    stdout=subprocess.PIPE,
+                                    stdin=subprocess.PIPE)
+            proc.communicate(input=self.host_behavior_data)
         super().tearDown()
 
-    def copy(self):
+    def copy(self, sdlba, blocks, slbs, **kwargs):
         """ Wrapper for nvme copy
             - Args:
-                - None
+                - sdlba : destination logical block address
+                - blocks : number of logical blocks (0-based)
+                - slbs : source range logical block address
+                - descriptor_format : copy descriptor format (optional)
+                - snsids : source namespace id (optional)
+                - sopts : source options (optional)
             - Returns:
-                - return code for nvme copy command.
+                - None
         """
-        copy_cmd = "nvme copy " + self.ctrl + \
-                   " --namespace-id=" + str(self.namespace) + \
-                   " --sdlba=" + str(self.start_block) + \
-                   " --blocks=" + str(self.range) + \
-                   " --slbs=" + str(self.range)
-        return self.exec_cmd(copy_cmd)
+        # skip if descriptor format not supported (default format is 0)
+        desc_format = kwargs.get("descriptor_format", 0)
+        if not self.ocfs & (1 << desc_format):
+            print(f"Skip copy because descriptor format {desc_format} is not supported")
+            return
+        # build copy command
+        copy_cmd = "nvme copy " + self.ns1 + \
+                   " --sdlba=" + str(sdlba) + \
+                   " --blocks=" + str(blocks) + \
+                   " --slbs=" + str(slbs)
+        if kwargs.get("descriptor_format") is not None:
+            copy_cmd += " --format=" + str(kwargs.get("descriptor_format"))
```

Comment:
`kwargs["descriptor_format"]`?

Reply:
You could also retrieve `descriptor_format` from the dict before testing and then just reuse its value if not `None` as follows.
```
descriptor_format = kwargs.get("descriptor_format")
if descriptor_format is not None:
    copy_cmd += f" --format={descriptor_format}"
```

Reply:
Yes, that is definitely more performant